internal/envelope/conn/envelope_conn.go

```diff
@@ -47,36 +49,44 @@ type EnvelopeHandler interface {
 // EnvelopeConn is the envelope side of the connection between a weavelet
 // and the envelope.
 type EnvelopeConn struct {
-	handler  EnvelopeHandler
-	conn     conn
-	metrics  metrics.Importer
-	weavelet *protos.WeaveletInfo
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+	conn      conn
+	metrics   metrics.Importer
+	weavelet  *protos.WeaveletInfo
+	running   errgroup.Group
+
+	mu  sync.Mutex // guards the fields below
+	err error      // error that stopped the envelope
 }
 
-// NewEnvelopeConn creates the envelope side of the connection between a
-// weavelet and an envelope. The connection uses (r,w) to carry messages.
-// Synthesized high-level events are passed to h.
+// NewEnvelopeConn creates the connection to an (already started) weavelet
+// and starts accepting messages from it. The connection uses (r,w) to carry
+// messages. Synthesized high-level events are passed to h.
 //
-// NewEnvelopeConn sends the provided protos.Weavelet to the weavelet.
-func NewEnvelopeConn(r io.ReadCloser, w io.WriteCloser, h EnvelopeHandler, wlet *protos.WeaveletSetupInfo) (*EnvelopeConn, error) {
+// The connection stops on error or when the provided context is canceled.
+func NewEnvelopeConn(ctx context.Context, r io.ReadCloser, w io.WriteCloser, wlet *protos.WeaveletSetupInfo) (*EnvelopeConn, error) {
+	ctx, cancel := context.WithCancel(ctx)
 	e := &EnvelopeConn{
-		handler: h,
-		conn:    conn{name: "envelope", reader: r, writer: w},
+		ctx:       ctx,
+		ctxCancel: cancel,
+		conn:      conn{name: "envelope", reader: r, writer: w},
 	}
 	// Send the setup information to the weavelet, and receive the weavelet
 	// information in return.
 	if err := e.send(&protos.EnvelopeMsg{WeaveletSetupInfo: wlet}); err != nil {
+		e.stop(err)
```

Comment:
Do we need to `stop` if we haven't called `Serve` yet?

Reply:
It's a bit of an overkill. But at the minimum, we should be calling e.conn.cleanup() here.

Reverted to that.